crates/driver/src/domain/competition/settled/mod.rs

```diff
@@ -0,0 +1,268 @@
+                            }
+                        };
+                        apply_factor(
+                            executed_in_surplus_token,
+                            match self.side {
+                                Side::Sell => max_volume_factor / (1.0 - max_volume_factor),
+                                Side::Buy => max_volume_factor / (1.0 + max_volume_factor),
+                            },
+                        )?
+                    },
+                )
+                .into(),
+            }),
+            order::FeePolicy::PriceImprovement {
+                factor: _,
+                max_volume_factor: _,
+                quote: _,
+            } => todo!(),
+            order::FeePolicy::Volume { factor } => Some(eth::Asset {
+                token: match self.side {
+                    Side::Sell => self.buy.token,
+                    Side::Buy => self.sell.token,
+                },
+                amount: {
+                    // Convert the executed amount to surplus token so it can be compared with
+                    // the surplus
+                    let executed_in_surplus_token = match self.side {
+                        Side::Sell => {
+                            self.executed.0 * self.prices.custom.sell / self.prices.custom.buy
+                        }
+                        Side::Buy => {
+                            self.executed.0 * self.prices.custom.buy / self.prices.custom.sell
+                        }
+                    };
+                    apply_factor(
+                        executed_in_surplus_token,
+                        match self.side {
+                            Side::Sell => factor / (1.0 - factor),
+                            Side::Buy => factor / (1.0 + factor),
+                        },
+                    )?
+                }
+                .into(),
+            }),
+        }
+    }
+
+    /// Protocol fee is defined by fee policies attached to the order.
+    ///
+    /// Denominated in NATIVE token
+    fn native_protocol_fee(
+        &self,
+        prices: &HashMap<eth::TokenAddress, auction::NormalizedPrice>,
+    ) -> Option<eth::TokenAmount> {
+        big_rational_to_u256(
+            &(self.protocol_fee()?.amount.0.to_big_rational() * self.surplus_token_price(prices).0),
+        )
+        .map(Into::into)
+        .ok()
+    }
+
+    /// Returns the normalized price of the trade surplus token
+    fn surplus_token_price(
+        &self,
+        prices: &HashMap<eth::TokenAddress, auction::NormalizedPrice>,
+    ) -> auction::NormalizedPrice {
+        match self.side {
+            Side::Buy => prices[&self.sell.token].clone(),
+            Side::Sell => prices[&self.buy.token].clone(),
+        }
+    }
+}
+
+fn apply_factor(amount: eth::U256, factor: f64) -> Option<eth::U256> {
+    Some(
+        amount.checked_mul(eth::U256::from_f64_lossy(factor * 1000000000000000000.))?
```

Comment:
You could multiply using `BigRational` numbers. That way you should have accurate results without having to multiply the factor with a huge number.
Since I see you defining that function twice it might be a good idea to have that as a general helper function in our `number` crate.